src/Forms/GridField/GridFieldDetailForm_ItemRequest.php

```diff
@@ -561,37 +561,115 @@ public function getEditLink($id)
             $id
         );
 
-        return $this->getStateManager()->addStateToURL($this->gridField, $link);
+        return $this->gridField->addAllStateToUrl($link);
     }
 
     /**
-     * @param int $offset The offset from the current record
-     * @return int|bool
+     * Return array of GridField items on current page plus
+     * first item on the next page and last item on the previous page
      */
-    private function getAdjacentRecordID($offset)
+    private function getGridFieldItemAdjacencies(): array
     {
-        $gridField = $this->getGridField();
-        $list = $gridField->getManipulatedList();
-        $state = $gridField->getState(false);
-        $gridStateStr = $this->getStateManager()->getStateFromRequest($this->gridField, $this->getRequest());
-        if (!empty($gridStateStr)) {
-            $state->setValue($gridStateStr);
-        }
-        $data = $state->getData();
-        $paginator = $data->getData('GridFieldPaginator');
+        $list = $this->getGridField()->getManipulatedList();
+        $paginator = $this->getGridFieldPaginatorState();
         if (!$paginator) {
-            return false;
+            return [];
         }
-
         $currentPage = $paginator->getData('currentPage');
         $itemsPerPage = $paginator->getData('itemsPerPage');
 
         $limit = $itemsPerPage + 2;
         $limitOffset = max(0, $itemsPerPage * ($currentPage-1) -1);
 
-        $map = $list->limit($limit, $limitOffset)->column('ID');
+        return $list->limit($limit, $limitOffset)->column('ID');
+    }
+
+    /**
+     * Get the current paginator state
+     */
+    private function getGridFieldPaginatorState(): GridState_Data
+    {
+        $state = $this->getGridField()->getState(false);
+        $gridStateStr = $this->getStateManager()->getStateFromRequest($this->gridField, $this->getRequest());
+        if (!empty($gridStateStr)) {
+            $state->setValue($gridStateStr);
+        }
+
+        return $state->getData()->getData('GridFieldPaginator');
+    }
+
+    /**
+     * Get the grid state for an adjacent record
+     */
+    private function getGridStateForAdjacentRecord(int $offset): GridState_Data
+    {
+        $gridField = $this->getGridField();
+        $map = $this->getGridFieldItemAdjacencies();
+        if (empty($map)) {
+            throw new LogicException('No adjacent records exist');
+        }
+
+        $state = clone $gridField->getState();
```

Comment:
I've just realised that this isn't cloning the underlying state _data_ - can you please add this method to the `GridState_Data` class?

```php


    public function __clone()
    {
        $data = new ArrayObject($this->data);
        $this->data = $data->getArrayCopy();
    }
```

Without this, the original state is still changed instead of only changing the _clone's_ data.

Reply:
Sorry, that implementation is flawed - this should work:

```php
public function __clone()
{
    $data = $this->toArray();
    $this->data = $data;
}
```

Reply:
DONE